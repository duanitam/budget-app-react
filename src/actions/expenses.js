import uuid from "uuid";
import database from '../firebase/firebase'
// ------- Action generators - Expenses -------

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState)=>{
        const uid = getState().auth.uid;
        const {  description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expense = { description, note, amount, createdAt};

        // Push the data in firebase database.
        // Return a promise so we can chain it -----
        return database.ref(`users/${uid}/expenses`).push(expense)
            .then( (ref) => {
            dispatch(addExpense({
                id:ref.key, ...expense})) })
            .catch((e) => {console.log(e)})
    }
};

export const removeExpense = ( { id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const editExpense = (id, update) => ({
    type:'EDIT_EXPENSE',
    id,
    update
});

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});


export const startSetExpenses = () => {

    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const expenses = [];

        return database.ref(`users/${uid}/expenses`).once('value')
            .then ((snapshot) => {
            snapshot.forEach( (item) => {
                expenses.push({id:item.key , ...item.val()})});
            dispatch(setExpenses(expenses));
        });
    }
};

export const startRemoveExpense = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}));
        })
    }
};

export const startEditExpense = (id, update) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(update)
            .then( () =>{
                dispatch(editExpense(id,update));
            })
    }
};