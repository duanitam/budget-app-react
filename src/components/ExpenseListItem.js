import React from 'react';
import { Link } from 'react-router-dom'

const ExpenseListItem = (props) => {
    return(
        <div>
        <Link to={`/edit/${props.id}`}>
            <h3>Item Description: {props.description}</h3>
        </Link>
            <br/>
            Note: {props.note} <br/>
            Amount: {props.amount}
            <br/><br/>
        </div>
    );
};

export default ExpenseListItem;