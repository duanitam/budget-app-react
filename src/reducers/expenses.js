
// Expenses Reducer
const expensesReducerDefaultState = [];
const expensesReducer = (state=expensesReducerDefaultState, action) => {

    switch (action.type){
        case 'ADD_EXPENSE':
            return ([...state, action.expense]);

        case 'REMOVE_EXPENSE':
            return state.filter(item => item.id !== action.id);

        case 'EDIT_EXPENSE':
            return state.map(item => (item.id === action.id) ? ({...item,...action.update}): (item));
        default:
            return state
    }
};

export default expensesReducer;