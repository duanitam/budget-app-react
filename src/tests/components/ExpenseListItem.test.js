import ExpenseListItem from "../../components/ExpenseListItem";
import {expenses} from "../fixtures/expenses";
import React from 'react';
import {shallow} from 'enzyme';

test('Should render expense item list', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});