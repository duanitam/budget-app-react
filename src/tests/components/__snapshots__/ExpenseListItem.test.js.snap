// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense item list 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Item Description: 
      Gum
    </h3>
  </Link>
  <br />
  Note: 
   
  <br />
  Amount: 
  195
  <br />
  <br />
</div>
`;
